lookout/style/format/rules.py

```diff
@@ -390,11 +404,69 @@ def _compile_rules(cls, rules: Sequence[Rule]) -> CompiledRulesType:
                 tuple(cls.CompiledNegatedRules(*v) for v in vr))
         return compiled_attrs
 
+    @staticmethod
+    def _get_parents(tree: DecisionTreeClassifier) -> Dict[int, int]:
+            parents = {x: i for i, x in enumerate(tree.children_left) if x != Tree.TREE_LEAF}
+            parents.update(
+                {x: i for i, x in enumerate(tree.children_right) if x != Tree.TREE_LEAF})
+            return parents
+
     @classmethod
-    def _prune_branches(cls, rules: Sequence[Rule],
-                        X: numpy.ndarray, Y: numpy.ndarray) -> List[Rule]:
-        # TODO(vmarkovtsev): implement this function
-        return rules
+    def reduced_error_prune(cls, model: DecisionTreeClassifier, validation_X, validation_y,
+                            possible_step_score_drop=0, possible_max_score_drop=0, verbose=False):
+        def _prune_tree(tree, node_to_prune):
+            backup = (tree.children_left[node_to_prune],
+                      tree.children_right[node_to_prune],
+                      tree.feature[node_to_prune])
+            tree.children_left[node_to_prune] = Tree.TREE_LEAF
+            tree.children_right[node_to_prune] = Tree.TREE_LEAF
+            tree.feature[node_to_prune] = Tree.TREE_UNDEFINED
+            return backup
+
+        def _restore_tree(tree, node, backup):
+            tree.children_left[node], tree.children_right[node], tree.feature[node] = backup
+
+        model = deepcopy(model)
+        tree = model.tree_
+        init_score = curent_score = model.score(validation_X, validation_y)
+        changes = True
+        bad_change = set()
+        parents = cls._get_parents(tree)
+        leafs = list(numpy.where(tree.children_left == Tree.TREE_LEAF)[0])
+        while changes:
+            changes = False
+            for leaf_index, leaf in enumerate(leafs):
+                if leaf not in parents:
+                    continue
+                parent = parents[leaf]
+                if parent in bad_change:
+                    continue
+                second_leaf = tree.children_right[parent]
+                second_leaf = second_leaf if second_leaf != leaf else tree.children_left[parent]
+                if tree.children_left[second_leaf] != Tree.TREE_LEAF or \
+                        tree.children_right[second_leaf] != Tree.TREE_LEAF:
+                    continue
+
+                backup = _prune_tree(tree, parent)
+                new_score = model.score(validation_X, validation_y)
+                score_change = (new_score - curent_score) / curent_score
+                full_score_change = (new_score - init_score) / init_score
```

Comment:
It should be possible to compute the score change from `tree.n_node_samples` and `tree.value` which would avoid another pass over the data (and keeping this way to perform a test of the validity of the faster method would be great!).

Reply:
yeah, it is true. I think to optimize this part, and I decided not to do it now. Just I case we evaluate our pruning methods and exclude this one, no reason for that. So I leave it as is for now. 

But if you think it is better to do it now, I can

Reply:
yup I think it might be necessary just because we can't easily test it on what we'll use it on without this optimization and it will be interesting to see if this method performs better than the greedy one on "real" data (big repos)